File: stickychat-api/src/main/java/com/dumbdogdiner/stickychat/api/mail/Letter.java
/*
 * Copyright (c) 2020-2021 Skye Elliot. All rights reserved.
 * Licensed under the GNU General Public License v3, see LICENSE for more information...
 */
package com.dumbdogdiner.stickychat.api.mail;

import java.util.UUID;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.BookMeta;

public interface Letter {
    /**
     * The unique ID of the player who sent this letter.
     *
     * @return {@link UUID}
     */
    UUID getSenderUniqueId();

    /**
     * The name of the player who sent this letter.
     *
     * @return {@link String}
     */
    String getSenderName();

    /**
     * The unique ID of the player who is the recipient of this letter.
     *
     * @return {@link UUID}
     */
    UUID getRecipientUniqueId();

    /**
     * The name of the player who is the recipient of this letter.
     *
     * @return {@link String}
     */
    String getRecipientName();

    /**
     * Get the content of this letter.
     *
     * @return A {@link String} containing the content.
     */
    String getContent();

    // TODO: convert to book
    BookMeta getMetadata();

    /**
     * Convert this letter into a book.
     *
     * @return An {@link ItemStack} of type book.
     */
    default ItemStack toItemStack() {
        ItemStack item = new ItemStack(Material.WRITTEN_BOOK);
        item.setItemMeta(this.getMetadata());
        return item;
    }
}

File: bukkit/src/main/kotlin/com/dumbdogdiner/stickychatbukkit/data/Letter.kt
package com.dumbdogdiner.stickychatbukkit.data

import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.BookMeta

/**
 * Represents a sent mail message.
 */
class Letter(
    val fromUuid: String,
    val fromName: String,
    val toUuid: String?,
    val toName: String,
    val title: String,
    val pages: List<String>,
    val createdAt: Long
) {
    fun asItem(): ItemStack {
        val item = ItemStack(Material.WRITTEN_BOOK, 1)
        val meta = item.itemMeta!! as BookMeta

        meta.title = title
        meta.author = fromName
        meta.pages = pages

        item.itemMeta = meta
        return item
    }
}
